### PR TITLE
Send remote access headers for CCS

### PR DESCRIPTION
**Pull request by n1v0lg**
This PR adds remote access headers to CCS requests sent by a querying cluster, if the new remote cluster security model is configured for the target fulfilling cluster.

This is done by extending `SecurityServerTransportInterceptor` to check if a given request requires remote access headers to be sent; if so, we fetch the remote index privileges for the authenticated subject associated with the request, construct the remote access headers (for the cluster credential and the remote access privileges of the user), and send these as transport headers. If we send remote access headers, we omit the standard `_xpack_security_authentication` transport header.

There are several conditions a request must meet to be sent with remote access headers: the connection must be to a remote cluster, and there must be a credential configured via cluster settings for the target cluster. A few other conditions (e.g., that the authentication type is supported) are there temporarily; we will lift these as we build out support. 

Since the fulfilling cluster logic for handling remote access headers is not implemented yet, CCS for feature-flagged deployments with configured cluster credentials will no longer work. This is safe since CCS will continue to fall back on the legacy security model unless the feature flag is set *and* the cluster credential setting is configured (which requires an active change to the settings).

### CODE REVIEW

**Comment by ywangd** [comment]
I'd prefer we start by just updating the existing `SecurityServerTransportInterceptor`. Like you said, for sending request, the new function has a dependency on the existing one. So it is better they stay close. If you still prefer the composite pattern, we can have it inside the existing class, e.g. something like:
```
public AsyncSender interceptSender(AsyncSender sender) {
   interceptForRemoteCluster(interceptForLocalCluster(sender))
}
```
I think it does the same job while hiding it better.

Also, for intercepting handler, I think we do want to update existing class (likely `ServerTransportFilter`) since it would otherwise need the existing class to bypass the authentication which feels risky. If this is the case, updating existing class for sending request feels more consistent.

Bottom line is that we can always refactor this internal implementation in future. But starting with existing class feels more straightforward and safer.

**Comment by n1v0lg** [comment]
@ywangd I'm good with moving this close to `SecurityServerTransportInterceptor` and refactoring later.

I do think we have a tendency at ES towards classes that are too big, but in this case `SecurityServerTransportInterceptor` stays manageable even with the remote access logic. We can always still have a separate class but once we move it out, I think it will look a lot like a standalone interceptor and might be surprising why we're not treating it as such.

I did start with s.th. like `interceptForRemoteCluster(interceptForLocalCluster(sender))` -- my reservation around this is that it essentially mimics the composite interceptor pattern but not idiomatically -- interceptors have the concept of chaining and composition built into them. To me, the question of how interceptors are composed is a more a question of "configuration" at the module level, rather than an internal implementation detail of `SecurityServerTransportInterceptor`. As you point out, inlining this in `SecurityServerTransportInterceptor` achieves the same functionality as using the composite pattern but does not use the pattern explicitly which feels a bit off to me.

That said, I'm for sure up for inlining for now as it is indeed more straightforward (and does not require exposing the `CompositeTransportInterceptor` class), and seeing if we can get to a good composite pattern later.

**Comment by ywangd** [comment]
Have you considered how interceptHandler should be implemented if it is a separate interceptor? I feel it could be rather inconvenient which is a major reason for me to suggest updating existing class. Maybe you have a better idea since I haven't put a lot time thinking about it yet.

**Comment by n1v0lg** [comment]
How intercepting incoming RCS 2.0 requests works is admittedly still a bit vague for me. 

Intercepting sending vs. receiving feel like orthogonal concerns so it might still fly to have a separate interceptor that handles sending, but include directly modify `SecurityServerTransportInterceptor` to handle receiving -- I agree, this feels inconsistent though.

I think you're right that it makes sense to start with the sender interceptor logic inlined, and then revisit moving things into a separate interceptor once we're tackling the receiver. I think it would be neat if we find a good structure there since it intuitively feels like we're composing interceptors. However, it requires some fiddling and for me to better understand how the fulfilling cluster will handle incoming requests (I have a decent idea around how authc and authz will work, but not yet how the new port and underlying transport profile factors in).

File: x-pack/plugin/security/qa/security-trial/src/javaRestTest/java/org/elasticsearch/xpack/security/remoteaccess/RemoteAccessHeadersForCcsRestIT.java [hunk 1]
```diff
@@ -0,0 +1,506 @@
...
+        if (false == useProxyMode) {
+            expectedActions.add(ClusterStateAction.NAME);
+        }
+        assertThat(
+            actualActionsWithHeaders.stream().map(CapturedActionWithHeaders::action).collect(Collectors.toUnmodifiableSet()),
+            equalTo(expectedActions)
+        );
+        for (CapturedActionWithHeaders actual : actualActionsWithHeaders) {
+            switch (actual.action) {
+                // the cluster state action is run by the system user, so we expect an authentication header, instead of remote access
```
Comment (n1v0lg) [inline x-pack/plugin/security/qa/security-trial/src/javaRestTest/java/org/elasticsearch/xpack/security/remoteaccess/RemoteAccessHeadersForCcsRestIT.java:344]: We will need figure out how to support this, but again due to scope, this will be a separate PR -- one where we tackle internal users.

File: x-pack/plugin/security/src/main/java/org/elasticsearch/xpack/security/transport/SecurityServerTransportInterceptor.java [hunk 2]
```diff
@@ -140,6 +194,108 @@ public <T extends TransportResponse> void sendRequest(
...
+                final Authentication authentication = securityContext.getAuthentication();
+                assert authentication != null : "authentication must be present in security context";
+                final Subject effectiveSubject = authentication.getEffectiveSubject();
+                final Optional<String> remoteClusterAlias = remoteClusterAliasResolver.apply(connection);
+                return remoteClusterAlias.isPresent()
+                    && remoteClusterAuthorizationResolver.resolveAuthorization(remoteClusterAlias.get()) != null
+                    && isWhitelistedForRemoteAccessHeaders(request)
+                    && effectiveSubject.getType().equals(Subject.Type.USER)
+                    && false == User.isInternal(effectiveSubject.getUser())
+                    && Arrays.stream(effectiveSubject.getUser().roles()).noneMatch(ReservedRolesStore::isReserved);
```
Comment (n1v0lg) [inline x-pack/plugin/security/src/main/java/org/elasticsearch/xpack/security/transport/SecurityServerTransportInterceptor.java:229]: This may be overly conservative but since working out how we will deal with reserved roles is still work-in-progress, it feels cleaner to explicitly fall back on the legacy model here.

File: x-pack/plugin/security/src/main/java/org/elasticsearch/xpack/security/transport/SecurityServerTransportInterceptor.java [hunk 3]
```diff
@@ -140,6 +194,108 @@ public <T extends TransportResponse> void sendRequest(
...
+            private void writeCredentialForClusterToContext(final String remoteClusterAlias, final ThreadContext threadContext) {
+                final String clusterCredential = remoteClusterAuthorizationResolver.resolveAuthorization(remoteClusterAlias);
+                // This can happen if the cluster credential was updated after we made the initial check to send remote access headers
+                // In this case, fail the request. In the future we may want to retry instead, to pick up the updated remote cluster
+                // configuration
+                if (clusterCredential == null) {
+                    throw new IllegalStateException(
+                        "remote cluster credential unavailable for target cluster [" + remoteClusterAlias + "]"
+                    );
+                }
```
Comment (ywangd) [inline x-pack/plugin/security/src/main/java/org/elasticsearch/xpack/security/transport/SecurityServerTransportInterceptor.java:289]: I wonder whether we should replace `boolean shouldSendWithRemoteAccessHeaders(...)` with something like
`Optional<RemoteAccessHeaders> getRemoteAccessHeaders(...)`. It avoids this problem and remove other duplicate reads as well. I personally also mostly prefer more elaborative code instead of a single large conditional statement, i.e. multiple `if` checks similar to [this](https://github.com/elastic/elasticsearch/blob/9d605b576f47c39ce7d7501a29181849eda6b713/x-pack/plugin/security/src/main/java/org/elasticsearch/xpack/security/authz/RBACEngine.java#L393). It also provides more opportunity for logging.

File: x-pack/plugin/security/qa/security-trial/src/javaRestTest/java/org/elasticsearch/xpack/security/remoteaccess/RemoteAccessHeadersForCcsRestIT.java [hunk 4]
```diff
@@ -0,0 +1,506 @@
...
+
+    @Override
+    public void tearDown() throws Exception {
+        super.tearDown();
+        ThreadPool.terminate(threadPool, 10, TimeUnit.SECONDS);
+    }
+
+    public void testRemoteAccessHeadersSentSingleRemote() throws Exception {
+        final BlockingQueue<CapturedActionWithHeaders> capturedHeaders = ConcurrentCollections.newBlockingQueue();
+        try (MockTransportService remoteTransport = startTransport("remoteNodeA", threadPool, capturedHeaders)) {
```
Comment (ywangd) [inline x-pack/plugin/security/qa/security-trial/src/javaRestTest/java/org/elasticsearch/xpack/security/remoteaccess/RemoteAccessHeadersForCcsRestIT.java:145]: TIL. This is a clever way for cross cluster testing.

**Review by ywangd — COMMENTED** [review]
I need a bit more time to read through the tests and ponder on the details (I don't see any real issues so far). In the meantime, here are some minor comments to get things rolling. Thanks!

File: x-pack/plugin/security/src/main/java/org/elasticsearch/xpack/security/transport/SecurityServerTransportInterceptor.java [hunk 5]
```diff
@@ -140,6 +194,108 @@ public <T extends TransportResponse> void sendRequest(
...
+                final TransportResponseHandler<T> handler
+            ) {
+                logger.debug("Sending request with remote access headers for [{}] action", action);
+
+                if (connection.getVersion().before(VERSION_REMOTE_ACCESS_HEADERS)) {
+                    handler.handleException(
+                        new TransportException(
+                            "Settings for remote cluster indicate remote access headers should be sent but target cluster version ["
+                                + connection.getVersion()
+                                + "] does not support receiving them"
```
[the remark this answers is not in the document]
Comment (n1v0lg) [reply x-pack/plugin/security/src/main/java/org/elasticsearch/xpack/security/transport/SecurityServerTransportInterceptor.java:246]: Yup, exactly my thinking -- throwing surfaces the configuration error explicitly. 

To pass it to the handler, it needs to be a `TransportException` (or subclass thereof). I can wrap it in a `SendRequestTransportException` here, so something like `new SendRequestTransportException(..., new IllegalArgumentException(...))`.

File: x-pack/plugin/security/src/main/java/org/elasticsearch/xpack/security/transport/SecurityServerTransportInterceptor.java (hunk 2 above)
Comment (n1v0lg) [reply x-pack/plugin/security/src/main/java/org/elasticsearch/xpack/security/transport/SecurityServerTransportInterceptor.java:229]: Yup, I'm fine with skipping this check -- my consideration was that a user with `superuser` role should always work but it doesn't matter much since this only applies to configured and feature flagged deployments either way.

File: x-pack/plugin/security/src/main/java/org/elasticsearch/xpack/security/transport/SecurityServerTransportInterceptor.java (hunk 3 above)
Comment (n1v0lg) [reply x-pack/plugin/security/src/main/java/org/elasticsearch/xpack/security/transport/SecurityServerTransportInterceptor.java:289]: `Optional<RemoteAccessHeaders> getRemoteAccessHeaders(...)` works and potentially simplifies things. I went with a boolean to keep the `shouldSendWithRemoteAccessHeaders` as single-purpose as possible but I don't feel strongly here.

Even if we go with `Optional<RemoteAccessHeaders> getRemoteAccessHeaders(...)` though, I wonder if we should still re-resolve the credential just before sending -- there is some time between the initial check and getting to the point of issuing the request (i.e., for resolving the roles etc). Resolving the credential again allows us to pick up last minute changes to the setting that would otherwise go undetected. We still don't get a proper lock on the setting though, so I can also see the argument for the simplified flow, where we only fetch it once. WDYT?

Comment (ywangd) [reply x-pack/plugin/security/src/main/java/org/elasticsearch/xpack/security/transport/SecurityServerTransportInterceptor.java:289]: I prefer fetching it once. I feel a proper solution is somewhere else, e.g. let `Connection` carries it alongside `clusterAlias`. So let's keep it simple here.

File: x-pack/plugin/security/src/main/java/org/elasticsearch/xpack/security/transport/SecurityServerTransportInterceptor.java (hunk 5 above)
[the remark this answers is not in the document]
Comment (ywangd) [reply x-pack/plugin/security/src/main/java/org/elasticsearch/xpack/security/transport/SecurityServerTransportInterceptor.java:246]: Can we just throw the exception? My understanding is that TransportService does the wrapping anyway (or if we consider this is composed inside another interceptor, it will be wrapped by `sendWithUser`). Directly calling `handler.handleException` is useful when we need proactively restore threadContext before handling exception. But in this case, we haven't modified threadContext yet.

[the remark this answers is not in the document]
Comment (n1v0lg) [reply x-pack/plugin/security/src/main/java/org/elasticsearch/xpack/security/transport/SecurityServerTransportInterceptor.java:246]: Ah yes for sure, we can throw. My thinking around passing the exception to the handler is still from back when this logic was in a standalone interceptor. But since this is not an internal implementation detail of the overall sec server interceptor it makes even more sense to throw here and let the exception get wrapped upstream.

File: x-pack/plugin/security/src/main/java/org/elasticsearch/xpack/security/transport/SecurityServerTransportInterceptor.java [hunk 6]
```diff
@@ -137,6 +190,142 @@ public <T extends TransportResponse> void sendRequest(
...
+            }
+
+            private boolean isAllowlistedForRemoteAccessHeaders(final TransportRequest request) {
+                return request instanceof ShardSearchRequest
+                    || request instanceof ShardFetchRequest
+                    || request instanceof SearchRequest
+                    || request instanceof ClusterSearchShardsRequest;
+            }
+
+            record RemoteAccessCredentials(String clusterAlias, String credentials) {
```
Comment (n1v0lg) [inline x-pack/plugin/security/src/main/java/org/elasticsearch/xpack/security/transport/SecurityServerTransportInterceptor.java:320]: This ended up a pretty neat abstraction (cheers for the suggestion Yang!) -- I think it makes sense to pull this up even further, to `RemoteClusterAuthorizationResolver` but I think that refactor makes more sense in a follow up PR to keep the current PR more focused.

File: x-pack/plugin/security/src/main/java/org/elasticsearch/xpack/security/transport/SecurityServerTransportInterceptor.java [hunk 7]
```diff
@@ -140,6 +195,108 @@ public <T extends TransportResponse> void sendRequest(
...
+                            writeCredentialForClusterToContext(remoteClusterAlias.get(), threadContext);
+                            new RemoteAccessAuthentication(authentication, roleDescriptorsIntersection).writeToContext(threadContext);
+                            sender.sendRequest(connection, action, request, options, contextRestoreHandler);
+                        }
+                    }, e -> contextRestoreHandler.handleException(new SendRequestTransportException(connection.getNode(), action, e)))
+                );
+            }
+
+            private boolean isWhitelistedForRemoteAccessHeaders(final TransportRequest request) {
+                return request instanceof ShardSearchRequest
```
[the remark this answers is not in the document]
Comment (jakelandis) [reply x-pack/plugin/security/src/main/java/org/elasticsearch/xpack/security/transport/SecurityServerTransportInterceptor.java:314]: I think the action names in org.elasticsearch.action.search.SearchTransportService may be the definitive list of actions we want to allow.

File: x-pack/plugin/security/qa/security-trial/src/javaRestTest/java/org/elasticsearch/xpack/security/remoteaccess/RemoteAccessHeadersForCcsRestIT.java (hunk 4 above)
Comment (jakelandis) [reply x-pack/plugin/security/qa/security-trial/src/javaRestTest/java/org/elasticsearch/xpack/security/remoteaccess/RemoteAccessHeadersForCcsRestIT.java:145]: ++

File: x-pack/plugin/security/src/main/java/org/elasticsearch/xpack/security/transport/SecurityServerTransportInterceptor.java (hunk 7 above)
[the remark this answers is not in the document]
Comment (n1v0lg) [reply x-pack/plugin/security/src/main/java/org/elasticsearch/xpack/security/transport/SecurityServerTransportInterceptor.java:314]: Thanks for the pointer! I'm seeing a few more candidate request types there (particularly around scroll). Some of the corresponding classes are package private, so I'm considering switching the allowlist to be based on action name instead of request type. The list will be longer (many actions map to the same request type) but this approach may more conventional and explicit. Still 👀 this one, and working out how scroll contexts work...

**Review by ywangd — COMMENTED** [review]
LGTM

I am approving the PR because the overall workflow and steps are sound. There are however two follow-up issues I think we should work on sooner rather than later.

1. Handling internal users
2. Final allowlist of actions/requests for support CCS

In my opinion, both of them, espcially item 1, are of high priority. I had a proposal on how it could work in the comment below. Happy to discuss further. Though very related, process wise, they do not have to be part of this PR. So I think we can wrap up this one first.

**Review by ywangd — APPROVED** [review]

File: x-pack/plugin/security/src/main/java/org/elasticsearch/xpack/security/transport/SecurityServerTransportInterceptor.java (hunk 7 above)
[the remark this answers is not in the document]
Comment (n1v0lg) [reply x-pack/plugin/security/src/main/java/org/elasticsearch/xpack/security/transport/SecurityServerTransportInterceptor.java:314]: > Using action names is likely the way to go.

Yup! I will push a preliminary list of action names to this PR and create an issue to track finalizing it. I'll go with a static set of all relevant actions which should give us a fast membership check.

> More specifically, all of them use the _system user to collect remote nodes (via GetClusterState cluster:monitor/state). Async search uses _async_search user to read and store partial results.

Absolutely. In sniff mode, even "regular" cross cluster search involves a GetClusterState action by the system user.

I agree that both support for internal users and the correct list of action names are high priority.

I like your suggested approach around handling internal users. There are a few caveats I want to think through here: 

I don't think we want to grant *all* access that a system user has within a cluster to remote cluster calls, and instead limit it to only the necessary actions (for the system user and barebones CCS it's only `cluster:monitor/state`). For example, the system user has `indices:data/write/*` access which we don't want to enable across clusters -- a malicious admin (or anyone else with access to the remote access API key credential) on a querying cluster could otherwise issue manually-crafted requests with a system user authentication instance in the header and escalate privileges.

There are three ways to go about this that I can think of: 

1. introduce new internal users (e.g. `remote_access_system_user`)
2. handle calls from existing internal users differently on the fulfilling cluster side, when processing a remote access request
3. enforce a privilege intersection with the remote access API key, and instruct administrators to manually include the privileges they require (for example, if an admin does not want to support async search, they don't have to include async user related privileges). This last option has more burden on the end-user but also gives them more control, and makes them explicitly aware of the full access scope a remote cluster has. As such, I don't want to dismiss this option just yet.

I'll put my thinking around the above in a Jira issue and get feedback there. Thanks for raising this, spot on as always!

File: x-pack/plugin/security/src/main/java/org/elasticsearch/xpack/security/transport/SecurityServerTransportInterceptor.java [hunk 8]
```diff
@@ -137,6 +190,142 @@ public <T extends TransportResponse> void sendRequest(
...
                     }
             }
+
+            private void assertNoRemoteAccessHeadersInContext() {
+                assert securityContext.getThreadContext().getHeader(REMOTE_ACCESS_CLUSTER_CREDENTIAL_HEADER_KEY) == null
+                    : "remote access headers should not be in security context";
+                assert securityContext.getThreadContext()
+                    .getHeader(RemoteAccessAuthentication.REMOTE_ACCESS_AUTHENTICATION_HEADER_KEY) == null
+                    : "remote access headers should not be in security context";
+            }
```
[the remark this answers is not in the document]
Comment (n1v0lg) [reply x-pack/plugin/security/src/main/java/org/elasticsearch/xpack/security/transport/SecurityServerTransportInterceptor.java:193]: I tried to scope things as tightly as possible to keep `SecurityServerTransportInterceptor` de-cluttered -- this method is only relevant in the context of `AsyncSender` and including it here (instead of on `SecurityServerTransportInterceptor`) excludes it from the list of methods listed for `SecurityServerTransportInterceptor` by the IDE (in Intellij it's CMD + F12)

File: x-pack/plugin/security/src/main/java/org/elasticsearch/xpack/security/transport/SecurityServerTransportInterceptor.java [hunk 9]
```diff
@@ -137,6 +190,142 @@ public <T extends TransportResponse> void sendRequest(
...
+                    || request instanceof ClusterSearchShardsRequest;
+            }
+
+            record RemoteAccessCredentials(String clusterAlias, String credentials) {
+                void writeToContext(final ThreadContext ctx) {
+                    ctx.putHeader(REMOTE_ACCESS_CLUSTER_CREDENTIAL_HEADER_KEY, encodedWithApiKeyPrefix(credentials));
+                }
+
+                private String encodedWithApiKeyPrefix(final String clusterCredential) {
+                    return Base64.getEncoder().encodeToString(("ApiKey " + clusterCredential).getBytes(StandardCharsets.UTF_8));
```
[the remark this answers is not in the document]
Comment (n1v0lg) [reply x-pack/plugin/security/src/main/java/org/elasticsearch/xpack/security/transport/SecurityServerTransportInterceptor.java:373]: I kept going back and forth on that -- encoding the whole header aligns with how we handle the other remote access header but it probably does make more sense to align it with the HTTP header instead. Will update 👍

File: x-pack/plugin/security/src/test/java/org/elasticsearch/xpack/security/transport/SecurityServerTransportInterceptorTests.java [hunk 10]
```diff
@@ -534,6 +563,294 @@ public boolean decRef() {
...
+            mock(SSLService.class),
+            securityContext,
+            new DestructiveOperations(
+                Settings.EMPTY,
+                new ClusterSettings(Settings.EMPTY, Collections.singleton(DestructiveOperations.REQUIRES_NAME_SETTING))
+            ),
+            remoteClusterAuthorizationResolver,
+            ignored -> Optional.of(remoteClusterAlias)
+        );
+        ClusterServiceUtils.setState(clusterService, clusterService.state()); // force state update to trigger listener
```
[the remark this answers is not in the document]
Comment (n1v0lg) [reply x-pack/plugin/security/src/test/java/org/elasticsearch/xpack/security/transport/SecurityServerTransportInterceptorTests.java:599]: Nope not necessary, it turns out. This was just copy&paste from a different test. Removing.

**Comment by n1v0lg** [comment]
@elasticsearchmachine run elasticsearch-ci/part-1

**Comment by n1v0lg** [comment]
@elasticsearchmachine run elasticsearch-ci/part-3
